doc/compiler/control/CheckpointRestoreCoordination.md

```diff
@@ -0,0 +1,434 @@
+<!--
+Copyright IBM Corp. and others 2024
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] https://openjdk.org/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+-->
+
+# Background
+
+This documention outlines how the compiler prepares for checkpoint and restore.
+This is a very dynamic process that requires the coordination of several
+components, as well as taking into account changes in the environment. This
+coordination is important for a few reasons:
+
+* The options can change on restore, and so the compiler threads cannot be in
+flight while the new options are being processed and applied.
+* The CPU restrictions can get lifted on restore, which can result low level CPU
+queries returning different answers within the same compilation.
+* The SCC may no longer exist on restore, or conversely, an SCC that did not
+exist pre-checkpoint now does exist.
+
+For all these reasons, and more, the compiler needs to ensure coordination to
+maintain coherency across a checkpoint/restore.
+
+# High Level Overview
+
+At bootstrap, the compiler registers callbacks for the
+`J9HOOK_VM_PREPARING_FOR_CHECKPOINT` and `J9HOOK_VM_PREPARING_FOR_RESTORE`
+hooks, namely `jitHookPrepareCheckpoint` and `jitHookPrepareRestore`
+respectively.
+
+When the application signals to the VM that a checkpoint should occur, the VM
+calls `jitHookPrepareCheckpoint` as part of the checkpoint process. This calls
+the top level method in the compiler that performs the necessary coordination:
+`TR::CompilationInfo::prepareForCheckpoint`.
+Once the coordination is complete, the executing thread returns to the VM, which
+eventually invokes CRIU to checkpoint itself.
+
+On restore, the VM calls `jitHookPrepareRestore` as part of the restore process.
+This first checks if further checkpointing is not allowed in order to remove the
+portability restrictions on the target CPU. It then calls the top level method
+in the compiler that performs the necessary coordination:
+`TR::CompilationInfo::prepareForRestore()`. Once the coordination is complete,
```

Comment:
`TR::CompilationInfo::prepareForRestore()` => `TR::CompilationInfo::prepareForRestore`. It looks like other methods are referred without `()`